plugin/src/main/kotlin/me/tylerbwong/gradle/metalava/Module.kt

```diff
@@ -13,6 +13,7 @@ internal sealed class Module {
 
     open val bootClasspath: Collection<File> = emptyList()
     abstract val compileClasspath: Collection<File>
+    open val taskVariants: Collection<String> = emptyList()
```

Comment:
Should we keep this specific to Android and just create an `internal` property instead? Could probably just keep the name `libraryVariants` then.

Reply:
Hi Tyler

Following-up on your comment about getting rid of

```  
    /**  
     * For Android modules defines which variant should be used to resolve classpath when Metalava  
     * generates or checks API.  
     */  
    var androidVariantName = "debug"
```

You're right, in that the change I've made should account for this.

However, attempting to implement that (and removing the above property), it ends up a, er, bit awkward.

Essentially, we're instantiating the Module at the start, and that has the information about the appropriate info associated with it, in it. Only after that instantiation do we know about the different `buildVariant`s, and those are only applicable to `Android`. So there are two issues:

- We must create the `Module` instance in order to find out about the _all_ `buildVariant`s
- `Android` wants to return N build variants, each of which will have a different classpath (the problem we're trying to resolve).

Thus, there is a mapping between "buildVariant -> classpath", but that isn't what we have on the other platforms, which are simply a single list. So one ends up with something like this:

```
class Android(private val extension: LibraryExtension) : Module() {  
    override val bootClasspath: Collection<File>  
        get() = extension.bootClasspath  
 
 override val compileClasspath: Collection<File>  
        get() = throw IllegalStateException("Unsupported for Android. Use compileClasspath(String).")  
 
    fun compileClasspath(libraryVariant: String): Collection<File> {  
        return extension.libraryVariants.find {  
 it.name.contains(libraryVariant, ignoreCase = true)  
        }?.getCompileClasspath(null)?.filter { it.exists() }?.files ?: emptyList()  
    }  
 
    val libraryVariants: Collection<String> // all the variants  
 get() = extension.libraryVariants.map { it.name }  
}
```

and that then has a knock-on effect on how we use it. Essentially, I can't see a 'nice' way of abstracting this such that it fits for `Android` AND the other `Module` types. So we end up with various casting and duplication in places, which isn't great really.

There are/were thoughts about perhaps getting rid of the `compileClasspath` property and just using the `compileClasspath(String)` function for all `Module` types; whether the parameter should be nullable and/or whether the signature/name can be coerced into something more 'generic' so that it fits all `Module`s. I couldn't square that, though.

Thoughts?